src/Database/LSMTree/Internal/Run/Reader.hs

```diff
@@ -0,0 +1,165 @@
+      !RawPage
+      !Word32
+      ![RawOverflowPage]
+
+-- | Refers to the underlying 'Run', so make sure it remains open while reading
+-- blobs from it!
+data BlobRef fhandle = BlobRef !(Run.Run fhandle) !BlobSpan
+
+-- | Stop using the 'Reader' after getting 'Empty'!
+next ::
+     HasFS IO h
+  -> HasBufFS IO h
+  -> Reader (FS.Handle h)
+  -> IO (Result (FS.Handle h))
+next fs bfs reader@Reader {..} = do
+    entryNo <- readPrimVar readerCurrentEntryNo
+    page <- readIORef readerCurrentPage
+    go entryNo page
+  where
+    go !entryNo !page =
+        -- take entry from current page (resolve blob if necessary)
+        case rawPageIndex page entryNo of
+          IndexNotPresent -> do
+            -- if it is past the last one, load a new page from disk, try again
+            readDiskPage fs bfs readerKOpsHandle >>= \case
+              Nothing -> do
+                close fs reader
+                return Empty
+              Just newPage -> do
+                writeIORef readerCurrentPage newPage
+                writePrimVar readerCurrentEntryNo 0
+                go 0 newPage  -- try again on the new page
+          IndexEntry key entry -> do
+            modifyPrimVar readerCurrentEntryNo (+1)
+            let entry' = fmap (BlobRef readerRun) entry
+            return (ReadSmallEntry key entry')
+          IndexEntryOverflow key entry lenSuffix -> do
+            -- TODO: we know that we need the next page, could already load?
+            modifyPrimVar readerCurrentEntryNo (+1)
+            let entry' = fmap (BlobRef readerRun) entry
+            overflowPages <- readOverflowPages fs bfs readerKOpsHandle lenSuffix
+            return (ReadLargeEntry key entry' page lenSuffix overflowPages)
+
+{-------------------------------------------------------------------------------
+  Utilities
+-------------------------------------------------------------------------------}
+
+-- | Returns 'Nothing' on EOF.
+readDiskPage :: HasFS IO h -> HasBufFS IO h -> FS.Handle h -> IO (Maybe RawPage)
+readDiskPage fs bfs h = do
+    mba <- newPinnedByteArray pageSize
+    handleJust (guard . FS.isFsErrorType FS.FsReachedEOF) (\_ -> pure Nothing) $ do
+      _ <- FS.hGetBufExactly fs bfs h mba 0 (fromIntegral pageSize)
+      ba <- unsafeFreezeByteArray mba
+      let !rawPage = unsafeMakeRawPage ba 0
+      return (Just rawPage)
+
+pageSize :: Int
+pageSize = 4096
+
+-- | Throws exception on EOF. If a suffix was expected, the file should have it.
+-- Reads full pages, despite the suffix only using part of the last page.
+readOverflowPages :: HasFS IO h -> HasBufFS IO h -> FS.Handle h -> Word32 -> IO [RawOverflowPage]
+readOverflowPages fs bfs h len = do
+    let lenPages = fromIntegral (roundUpToPageSize len)  -- always read whole pages
+    mba <- newPinnedByteArray lenPages
+    _ <- FS.hGetBufExactly fs bfs h mba 0 (fromIntegral lenPages)
+    ba <- unsafeFreezeByteArray mba
+    return $ rawBytesToOverflowPages (RB.fromByteArray 0 lenPages ba)
```

Comment:
It's possible, but I'd end up re-implementing most of `rawBytesToOverflowPages`. We could provide `byteArrayToOverflowPages` in the `RawOverflowPages` module, but it would also just be a wrapper for `rawBytesToOverflowPages` (the other direction doesn't work I think, unless there is an offset argument).

Reply:
Okay, but it might be wort using an unsafe version that doesn't perform the `isByteArrayPinned` check at run-time. In fact, that's what `rawBytesToOverflowPagesPinned` does, so maybe we can expose that and use it instead? We should probably also add an assertion for `isByteArrayPinned` in there to make sure we catch unsafe uses.